### PR TITLE
Robot

### PR DESCRIPTION
**Pull request by dietpop**
Output:
clm-mwaldorf:robot_name mwaldorf$ ruby robot_creator.rb
I'm your new robot named SU182, but you can call me Sparkles!
I'm your new robot named XU578, but you can call me Sparkles!
I'm your new robot named LG326, but you can call me Sparkles!
I'm your new robot named EX439, but you can call me Sparkles!
I'm your new robot named QY512, but you can call me Sparkles!
I'm your new robot named DG433, but you can call me Sparkles!
I'm your new robot named SC611, but you can call me Sparkles!
I'm your new robot named AS838, but you can call me Sparkles!
I'm your new robot named WS411, but you can call me Sparkles!
I'm your new robot named HY565, but you can call me Sparkles!
I'm your new robot named TK397, but you can call me Sparkles!
I'm your new robot named CI447, but you can call me Sparkles!
I'm your new robot named IS424, but you can call me Sparkles!
I'm your new robot named AA111, but you can call me Sparkles!
/Users/mwaldorf/level_up_exercises/robot_name/registry.rb:10:in `check_duplicate': AA111 is a duplicate. (NameCollisionError)

### CODE REVIEW

File: robot_name/name_collision_error.rb [hunk 1]
```diff
@@ -0,0 +1 @@
+class NameCollisionError < RuntimeError; end
```
Comment (paulh16) [inline robot_name/name_collision_error.rb]: Good job splitting up your errors in two separate classes.

File: robot_name/name_generator.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+class NameGenerator
+  def generate_name
+    [*1..5].inject("") do |name, i|
```
Comment (paulh16) [inline robot_name/name_generator.rb]: Really nice use of the split to generate an array of numbers. Consider a better name for the second block variable, `i`.

File: robot_name/name_generator.rb [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+      i < 3 ? name << [*'A'..'Z'].sample : name << rand(10).to_s
```
Comment (paulh16) [inline robot_name/name_generator.rb]: I'd personally have generate name take in two parameters for the number of characters and numbers for the name, with some default values. Then I'd use these parameters with the `times` method to generate the name with the required length.

Comment (paulh16) [inline robot_name/name_generator.rb]: I love the splat! `[*'A'..'Z']`

File: robot_name/name_generator.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+class NameGenerator
```
Comment (paulh16) [inline robot_name/name_generator.rb]: Really nice adherence to the Single Responsibility Principle.

File: robot_name/registry.rb [hunk 5]
```diff
@@ -0,0 +1,12 @@
+require './name_collision_error.rb'
```
Comment (paulh16) [inline robot_name/registry.rb]: You don't need the `.rb` at the end. You could alternatively use `require_relative 'name_collision_group`.

File: robot_name/registry.rb [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+class Registry
+  attr_accessor :names
+
+  def initialize
+    @names = []
+  end
+
+  def check_duplicate(name)
```
Comment (paulh16) [inline robot_name/registry.rb]: Consider a better name for this method, ie: `check_for_duplicate_name`.

File: robot_name/robot.rb [hunk 7]
```diff
@@ -0,0 +1,8 @@
+class Robot
+  attr_accessor :name
+
+  def initialize(name)
+    @name = name
+    puts "I'm your new robot named #{name}, but you can call me Sparkles!"
```
Comment (paulh16) [inline robot_name/robot.rb]: I'd personally extract this into a new private method so it doesn't clutter up your `initialize` method.

File: robot_name/robot_creator.rb [hunk 8]
```diff
@@ -0,0 +1,30 @@
+require './name_input_error.rb'
```
Comment (paulh16) [inline robot_name/robot_creator.rb]: See previous comment about omitting the `.rb` and the alternative use of `require_relative`.

File: robot_name/robot_creator.rb [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+require './registry.rb'
+require './robot.rb'
+require './name_generator.rb'
+
+class RobotCreator
+  REG_EXP_NAME =  /[[:alpha:]]{2}[[:digit:]]{3}/
```
Comment (paulh16) [inline robot_name/robot_creator.rb]: Really nice regex!

File: robot_name/robot_creator.rb [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+  NAME_INPUT_ERROR = "There was a problem with the user's robot name!"
+
+  def initialize
+    @robot_list = Registry.new
+  end
+
+  def create_robots(number_of_robots, *arg)
+    (0...number_of_robots).each do |i|
```
Comment (paulh16) [inline robot_name/robot_creator.rb]: Consider using the `.times` method, called on `number_of_robots`.

Comment (paulh16) [inline robot_name/robot_creator.rb]: I'd recommend a better name for `i`, such as `robot`.

File: robot_name/robot_creator.rb [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+      robot_name = arg[i] || NameGenerator.new.generate_name
+      validate_input_name(robot_name)
+      @robot_list.check_duplicate(robot_name)
+      @robot_list.names << Robot.new(robot_name).name
+    end
+  end
+
+  def validate_input_name(name)
```
Comment (paulh16) [inline robot_name/robot_creator.rb]: Consider using a private method here.

File: robot_name/robot_creator.rb [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+  REG_EXP_NAME =  /[[:alpha:]]{2}[[:digit:]]{3}/
+  NAME_INPUT_ERROR = "There was a problem with the user's robot name!"
+
+  def initialize
+    @robot_list = Registry.new
+  end
+
+  def create_robots(number_of_robots, *arg)
+    (0...number_of_robots).each do |i|
+      robot_name = arg[i] || NameGenerator.new.generate_name
```
Comment (paulh16) [inline robot_name/robot_creator.rb]: Excellent use of this idiom!

File: robot_name/robot_creator.rb [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+
+class RobotCreator
+  REG_EXP_NAME =  /[[:alpha:]]{2}[[:digit:]]{3}/
+  NAME_INPUT_ERROR = "There was a problem with the user's robot name!"
+
+  def initialize
+    @robot_list = Registry.new
+  end
+
+  def create_robots(number_of_robots, *arg)
```
Comment (paulh16) [inline robot_name/robot_creator.rb]: Consider renaming `arg` to better explain what it does.

File: robot_name/robot_creator.rb [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+    (0...number_of_robots).each do |i|
+      robot_name = arg[i] || NameGenerator.new.generate_name
+      validate_input_name(robot_name)
```
Comment (paulh16) [inline robot_name/robot_creator.rb]: You could always extract lines 17 and 18 into their own method, which validates the robot generally. But this is pretty nit-picky.

File: robot_name/robot_creator.rb [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+      @robot_list.check_duplicate(robot_name)
```
Comment (paulh16) [inline robot_name/robot_creator.rb]: Good OO design. You are asking the proper class, `Registry`, whether this is a duplicate.

File: robot_name/robot_creator.rb [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+      @robot_list.names << Robot.new(robot_name).name
```
Comment (paulh16) [inline robot_name/robot_creator.rb]: Consider extracting this line into a well-named method, such as `add_robot_to_registry`.

**Comment by paulh16** [comment]
@dietpop Fantastic job overall! Great overall class design and adherence to the Single Responsibility Principle. Most of my suggestions we in naming methods and variables, possible method extractions and using different `Enumerable` methods, such as `times`.

@djkotowski I recommend this PR as closed and good to go.

File: robot_name/name_generator.rb (hunk 3 above)
Comment (djkotowski) [inline robot_name/name_generator.rb]: Avoid the ternary operator here.  This is way too complex an expression to be readable as a ternary.

File: robot_name/registry.rb (hunk 5 above)
Comment (djkotowski) [inline robot_name/registry.rb]: Going to push "should" instead of "could" here.  `require_relative` is more secure for requiring files.

**Comment by djkotowski** [comment]
Good job.  Take a look at the comments, but need for resubmission.

**Pull request closed without merging**